File: src/api/static.js
const app = require("../app");
const logger = require("../logger");
const { get_by_schemaname_and_username } = require("../dao/schema");
const { find_all } = require("../dao/schema_screenshot");

app.get('/api/static/schema/:username/:schemaname', async function(req, res){
  logger.debug("GET /api/static/schema", req.params);

	const schema = await get_by_schemaname_and_username(req.params.schemaname, req.params.username);

	if (!schema){
		res.status(404).end();
		return;
	}

	const screenshots = await find_all(schema.id);

	if (!screenshots || screenshots.length == 0){
		res.status(404).end();
		return;
	}


	const html = /*html*/`
	<!DOCTYPE HTML>
	<html>
		<head>
			<meta name="og:title" content="${schema.name} by ${req.params.username}"/>
			<meta name="og:type" content="Schematic"/>
			<meta name="og:url" content="${process.env.BASE_URL}/#/schema/${req.params.username}/${schema.name}"/>
			<meta name="og:image" content="${process.env.BASE_URL}/api/schema/${schema.id}/screenshot/${screenshots[0].id}"/>
			<meta name="og:site_name" content="Block exchange"/>
			<meta name="og:description" content="${schema.description}"/>
			<meta http-equiv="refresh" content="0; url=${process.env.BASE_URL}/#/schema/${req.params.username}/${schema.name}" />
		</head>
		<body>
		</body>
	</html>
	`;

	res.header("Content-type", "text/html").send(html);
});

File: web/static.go
package web

import (
	"blockexchange/types"
	"html/template"
	"net/http"
	"os"

	"github.com/gorilla/mux"
)

const template_str = `
<!DOCTYPE HTML>
<html>
	<head>
		<meta name="og:title" content="{{.Schema.Name}} by {{.Username}}"/>
		<meta name="og:type" content="Schematic"/>
		<meta name="og:url" content="{{.BaseURL}}/#/schema/{{.Username}}/{{.Schema.Name}}"/>
		<meta name="og:image" content="{{.BaseURL}}/api/schema/{{.Schema.ID}}/screenshot/{{.Screenshot.ID}}"/>
		<meta name="og:site_name" content="Block exchange"/>
		<meta name="og:description" content="{{.Schema.Description}}"/>
		<meta http-equiv="refresh" content="0; url={{.BaseURL}}/#/schema/{{.Username}}/{{.Schema.Name}}" />
	</head>
	<body>
	</body>
</html>
`

var tmpl = template.Must(template.New("main").Parse(template_str))

type TemplateData struct {
	Schema     *types.SchemaSearchResult
	Screenshot *types.SchemaScreenshot
	Username   string
	BaseURL    string
}

func (api *Api) GetStaticView(w http.ResponseWriter, r *http.Request) {
	vars := mux.Vars(r)
	schema_name := vars["schema_name"]
	user_name := vars["user_name"]

	schema, err := api.SchemaSearchRepo.FindByUsernameAndSchemaname(schema_name, user_name)
	if err != nil {
		SendError(w, 500, err.Error())
		return
	}

	if schema == nil {
		SendError(w, 404, "Not found")
		return
	}

	screenshots, err := api.SchemaScreenshotRepo.GetBySchemaID(schema.ID)
	if err != nil {
		SendError(w, 500, err.Error())
		return
	}

	if screenshots == nil || len(screenshots) < 1 {
		SendError(w, 500, "no screenshots found")
		return
	}

	data := TemplateData{
		Schema:     schema,
		Screenshot: &screenshots[0],
		Username:   user_name,
		BaseURL:    os.Getenv("BASE_URL"),
	}

	w.Header().Set("Content-Type", "text/html")
	tmpl.Execute(w, data)
}
